### DISCUSSION TITLE
User login support to allow sharing annotations between desktop and Logseq

### DISCUSSION THREAD

**Discussion by VictorVow** [duydl/logseq-calibre-plugin#11] [General]
Currently without user login support I don't know how you can see highlights created in either desktop or Logseq simultaneously.

In Calibre > View > Browse annotations I can see all annotations and what user owns them (`Anonymous Content server user` for those made in Logseq, and `Local E-book viewer user`/`localuser` for those made in desktop Calibre). 

When I open the book in Logseq I see the former annotations, when I open the book in Logseq the latter annotations. Having user login support would made it so that I can use `localuser` to store/view/edit all annotations on either app.

**Comment by CobriMediaJulien** [comment]
+1 for user log in. Then u could use whatever viewer you want and sync annotations to logseq. Maybe without a pop but just a setting in the pluging setting for password and user?

@duydl is this hard to implement? i will have a look later but i´m new to coding ... :(

**Comment by duydl** [comment]
Implemented in newest version. Please try it and give feedback in this issue.

**Comment by VictorVow** [comment]
Thanks for adding the functionality, it doesn't seem to be working for me however. 

Restarted Logseq inbetween each test:
1) Calibre setting "Require username and password to access the Content server" disabled = can use your plugin to open books and access content server at http://localhost:3296, highlights utilise anonymous user
2) Calibre setting "Require username and password to access the Content server" disabled + added Calibre username + password in your plugin settings = can use your plugin to open books and access content server at http://localhost:3296, highlights utilise anonymous user
3) Calibre setting "Require username and password to access the Content server" **enabled** + added Calibre username + password in your plugin settings = using your plugin to open books instead shows a blank pane but with the floating draggable toolbar present

#3 in console log I can see this error "Failed to load resource: the server responded with a status of 401 (Unauthorized)". Not sure why this is the case, I can use the exact same username + password to access the content server through a web browser.

**Comment by duydl** [comment]
Hi thanks for the feedback. I am indeed having problem making the viewer working with login. However, could you confirm  if:
- Being able to add new book with slash command "Calibre Annotation"
- Being able to sync annotations made with **browser**.

**Comment by VictorVow** [comment]
> Being able to add new book with slash command "Calibre Annotation"

Yes, that works even with `Require username and password to access the Content server` and the correct username/password.

However it's not creating a separate page for the book, instead it adds the book as a new block. I'm not sure if this is a bug or new expected behaviour:

![](https://i.imgur.com/fOiJLHS.png)

> Being able to sync annotations made with **browser**.

This works too. 

By the way the  "Search Calibre library" side pane that opens when you try to add a Calibre book is incompatible with [AwesomeUI](https://github.com/yoyurec/logseq-awesome-ui). I've got [Tabs](https://github.com/pengx17/logseq-plugin-tabs) installed also but that seems to be compatible. Have had to disable it when trying to test the new version of this plugin, where before your plugin used a floating pane so it wasn't an issue:

![](https://i.imgur.com/9cXSwj3.png)

**Comment by duydl** [comment]
> However it's not creating a separate page for the book, instead it adds the book as a new block. I'm not sure if this is a bug or new expected behaviour:

This has always been an optional setting. Disable it by unchecking the box:
![image](https://github.com/user-attachments/assets/531fdd35-f450-4ddd-845e-37e70b99b3f8)

> incompatible with [AwesomeUI](https://github.com/yoyurec/logseq-awesome-ui).

Actually it is because a combination of both Tab and AwesomeUI. If you could type in the search bar and only the close button is overlapped, you could still exit with Escape button. A possible fix is to add some custom CSS like this
![image](https://github.com/user-attachments/assets/f346f7b3-f85a-4c16-acc1-91ef073389be)

**Comment by CobriMediaJulien** [comment]
I tried implementing your plugin to log into the calibre web server using a username and pw. Problem is to hand over pw and user name via the iframe. I tried several stuff. I was even able to log in. But couldnt refresh the iframe page so i was stuck in a loading bar. Right now i suggestion is to use the web ebook reader of calibre web and not the iframe. You could just disable the iframe and people can use the browser windows to read and make annotations.

you would have to let logseq as the master application handle the log in stuff via a security token and handle this to the iframe. that was when i gave up, tho calibre web server docu is sh****.

the problem is with the iframe, that with every query to the server u have to pass the username and pw. That is not possible. i think a iframe does not support this. so if u just hand it 1 time then you will be stuck in a loading bar because it tries to load the reader. But it will decline the query for the content. Easiest way to have a username is to just disable the iframe and use a webbrwoser, log into calibre and use the annotations there. sync still works.

**Comment by duydl** [comment]
@CobriMediaJulien Thank you for the suggestion and trying. I think Calibre Server is simple Digest Authen though the problem was with iframe and CORS and stuff. 
 
> You could just disable the iframe and people can use the browser windows to read and make annotations.

I have just implemented this in the newest version to allow opening viewer and annotation in browser. It opens multiple new tab though so not that convenient:

![image](https://github.com/user-attachments/assets/fe063803-a909-4f63-8de5-42935b6af07f)

**Comment by VictorVow** [comment]
> This has always been an optional setting. Disable it by unchecking the box: ![image](https://private-user-images.githubusercontent.com/56506156/352977318-531fdd35-f450-4ddd-845e-37e70b99b3f8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIyNTY4NzgsIm5iZiI6MTcyMjI1NjU3OCwicGF0aCI6Ii81NjUwNjE1Ni8zNTI5NzczMTgtNTMxZmRkMzUtZjQ1MC00ZGRkLTg0NWUtMzdlNzBiOTliM2Y4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI5VDEyMzYxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE0NTNjZmZmNTZhYmNhNTFmODU5NDBhNDU0NjBjZDY3OTU5OTkwMmM5NTUzNDg3NmYwODg4M2QzYzA4MDM2MjcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.mXczmgOQ-msm976RwkrQz9Kl9Cj45efI-nT4AZtYoFE)

Thanks, I'd overlooked that setting.

> Actually it is because a combination of both Tab and AwesomeUI. If you could type in the search bar and only the close button is overlapped, you could still exit with Escape button. A possible fix is to add some custom CSS like this ![image](https://private-user-images.githubusercontent.com/56506156/352980191-f346f7b3-f85a-4c16-acc1-91ef073389be.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIyNTY4NzgsIm5iZiI6MTcyMjI1NjU3OCwicGF0aCI6Ii81NjUwNjE1Ni8zNTI5ODAxOTEtZjM0NmY3YjMtZjg1YS00YzE2LWFjYzEtOTFlZjA3MzM4OWJlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzI5VDEyMzYxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThjMTExNWI2MTY5OGE1Y2Y0NTA3YzYzNjg3MTIyOGNhYTY5YzU0MzUxZmMwZjNlZjI5MTdiY2U3ZmE0NzVlZWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.INsCpEBILb4O6SSMmAiqu3eQyhIi5V47IpMrCAQmnfY)

Thanks again, that works splendidly.

**Comment by ArcaneSage1** [comment]
Is this issue with iframe and CORS etc due to a limitation with Logseq or with Calibre? I think a bug report should be posted for the relevant one as I'd like to use the viewer within Logseq rather than a separate browser window.

**Reply by duydl** [reply]
Hi. Sorry for the late reply. 
I would say it is a problem with approach of using iframe in general. Though could also be a limitation of Logseq as Logseq plugin doesn't expose webview of electron or other backends API, which more flexible with CORs than iframe, even bypassing it.
I believe there must be a way to authenticate with just iframe, but it is pretty niche and I haven't had time to look into it further.

**Reply by dag2108** [reply]
would love it if this could get implemented, any chance we can get this issue opened so it could get resolved eventually?

**Reply by duydl** [reply]
Hi. I moved this to discussion because user login support was added. There just new complication that the viewer no longer could be opened from Logseq when login required (but you could still open in browser and sync annotation into Logseq). People seem to  dont mind it much.
As it is technically new problem, feel free to create a separate issue for it. 

I little bit hesitate to spend much time on it as Logseq seem to be having major change in  plugin system, and there may be more effective way to bypass the CORS. Check it out here: https://discuss.logseq.com/t/a-survey-for-the-upcoming-plugin-system/773
